src/coreclr/jit/valuenum.cpp

```diff
@@ -7768,82 +7777,80 @@ void Compiler::fgValueNumberAssignment(GenTreeOp* tree)
+                        ValueNum storeVal  = rhsVNPair.GetLiberal();
+                        ValueNum newHeapVN = ValueNumStore::NoVN;
 
-                        // The final field in the sequence will need to match the 'indType'
+                        // We will check that the final field in the sequence matches 'indType'.
                         var_types indType = lhs->TypeGet();
-                        ValueNum  fldMapVN =
-                            vnStore->VNApplySelectors(VNK_Liberal, fgCurMemoryVN[GcHeap], firstFieldOnly);
-
-                        // The type of the field is "struct" if there are more fields in the sequence,
-                        // otherwise it is the type returned from VNApplySelectors above.
-                        var_types firstFieldType = vnStore->TypeOfVN(fldMapVN);
-
-                        // The value number from the rhs of the assignment
-                        ValueNum storeVal    = rhsVNPair.GetLiberal();
-                        ValueNum newFldMapVN = ValueNumStore::NoVN;
 
                         // when (obj != nullptr) we have an instance field, otherwise a static field
                         // when (staticOffset != nullptr) it represents a offset into a static or the call to
                         // Shared Static Base
                         if ((obj != nullptr) || (staticOffset != nullptr))
                         {
-                            ValueNum valAtAddr = fldMapVN;
-                            ValueNum normVal   = ValueNumStore::NoVN;
+                            var_types firstFieldType;
+                            ValueNum  firstFieldSelectorVN =
+                                vnStore->VNForFieldSelector(fldSeq->GetFieldHandle(), &firstFieldType);
 
+                            // Construct the "field map" VN. It represents memory state of the first field
+                            // of all objects on the heap. This is our primary map.
+                            ValueNum fldMapVN = vnStore->VNForMapSelect(VNK_Liberal, firstFieldType,
+                                                                        fgCurMemoryVN[GcHeap], firstFieldSelectorVN);
+
+                            ValueNum firstFieldValueSelectorVN = ValueNumStore::NoVN;
                             if (obj != nullptr)
                             {
-                                // Unpack, Norm,Exc for 'obj'
-                                ValueNum vnObjExcSet;
-                                vnStore->VNUnpackExc(obj->gtVNPair.GetLiberal(), &normVal, &vnObjExcSet);
-                                vnExcSet = vnStore->VNExcSetUnion(vnExcSet, vnObjExcSet);
-
-                                // construct the ValueNumber for 'fldMap at obj'
-                                valAtAddr = vnStore->VNForMapSelect(VNK_Liberal, firstFieldType, fldMapVN, normVal);
+                                firstFieldValueSelectorVN = vnStore->VNLiberalNormalValue(obj->gtVNPair);
```

Comment:
Do we not need to unpack the VN like before?

Reply:
Ah, scratch that, I see this just uses VNLiberalNormalValue now.

Reply:
Note that the tracking of exception sets this code did before, i. e.
```c++
vnExcSet = vnStore->VNExcSetUnion(vnExcSet, vnObjExcSet);
```
was dead code because nothing uses `vnExcSet` in this method - the `ASG` tree is assigned `VNForVoid()` at the end, [always](https://github.com/dotnet/runtime/blob/3c671454948a8b14242f0ed98ae12c0f6e3a96d1/src/coreclr/jit/valuenum.cpp#L7881), i. e. we were always losing all the exceptions. We are "ok" correctness-wise here only because CSE is the only consumer of exception sets right now and it does not look at trees marked with `GTF_ASG`.